app/crud/serviceMetric.py

```diff
@@ -1,19 +1,142 @@
+        metrics = (
+            self.db_session.query(ServiceMetric.metricId,ServiceMetric.value,ServiceMetric.timestamp)
+        )
+        query = self.db_session.query(ServiceMetric.metricId, ServiceMetric.value, ServiceMetric.timestamp)  # Query initialized here
+        if from_date and to_date:
+          query = query.filter(ServiceMetric.timestamp >= from_date, ServiceMetric.timestamp <= to_date)
+        elif from_date:
+          query = query.filter(ServiceMetric.timestamp >= from_date)
+        elif to_date:
+          query = query.filter(ServiceMetric.timestamp <= to_date)
+        
+        query = query.filter(ServiceMetric.serviceId == service_id).order_by(ServiceMetric.timestamp.desc())
+        metrics = query.all()
+        return metrics
+     
+     
+      
+    def get_last_metrics(self, scorecard_id: int, service_id: int)-> list[ServiceMetric]:
+        
+        subquery = (
+        self.db_session.query(
+            ServiceMetric.metricId,
+            func.max(ServiceMetric.timestamp).label('latest_timestamp')
+        )
+        .filter(
+            ServiceMetric.serviceId == service_id,
+            ServiceMetric.metricId.in_(self.scorecardMetrics.getMetricByScoreCradId(scorecard_id))
+        )
+        .group_by(ServiceMetric.metricId)
+        .subquery()
+    )
+
+        latest_metrics = self.db_session.query(
+        ServiceMetric.metricId,
+        ServiceMetric.value,
+        ServiceMetric.timestamp
+     ).join(
+        subquery,
+        and_(
+            ServiceMetric.metricId == subquery.c.metricId,
+            ServiceMetric.timestamp == subquery.c.latest_timestamp
+        )
+     ).filter(
+        ServiceMetric.serviceId == service_id
+     ).all()
+
+   
+        print("servicemetric:", latest_metrics)
+        return latest_metrics
+     
+    
+    def get_timestamp(self, service_id: int, scorecard_id: int):
+        subquery = self.scorecardMetrics.getMetricByScoreCradId(scorecard_id)
+
+        update_time = self.db_session.query(ServiceMetric.timestamp)\
+            .filter(ServiceMetric.serviceId == service_id, ServiceMetric.metricId.in_(subquery))\
+            .order_by(ServiceMetric.timestamp.desc())\
+            .first()
+
+        if update_time:
+            formatted_time = update_time.timestamp.strftime(
+                '%Y-%m-%d %H:%M:%S')
+            return formatted_time
+        else:
+            return None
+ 
+
+    def get_calculated_value(self, service_id: int, scorecard_id: int):
+        scorecard_metrics = self.scorecardMetrics.get_metric(scorecard_id)
+        print("Scorecard metrics:", scorecard_metrics)
+        metric_info_dict = {
+            metric.metricId: (
+                metric.criteria, metric.desiredValue, metric.weight)
+            for metric in scorecard_metrics
+        }
+
+        metric_info_dict = {}
+        for metric in scorecard_metrics:
+            metric_type = self.Metric.getMetricType(metric.metricId)
```

Comment:
query the DB once and get list of metrics corresponding to the list of IDs embedded in the `scorecard_metrics`. then extract the type information of each metric given its ID from the returned list.